### PR TITLE
Support negative null-checks when we offer a fix to use pattern-matching

### PR DESCRIPTION
**Pull request by alrz**
Follow up: https://github.com/dotnet/roslyn/pull/20240
- [x] Fix https://github.com/dotnet/roslyn/issues/21097
- [x] Fix https://github.com/dotnet/roslyn/issues/24286

### CODE REVIEW

**Comment by alrz** [comment]
@Pilchie this is actually IDE related. same for https://github.com/dotnet/roslyn/pull/20240, all of which blocked by #19907 awaiting your review.

**Comment by Pilchie** [comment]
Thanks and sorry!

Tagging the rest of @dotnet/roslyn-ide for review too.

**Comment by alrz** [comment]
Please hold off the review untill the linked PR goes in. Thanks.

**Comment by alrz** [comment]
@sharwell this is locally green, ready for review. thanks.

**Comment by alrz** [comment]
@sharwell any chance we can have this for 15.6?

File: src/EditorFeatures/CSharpTest/UsePatternMatching/CSharpAsAndNullCheckTests_FixAllTests.cs [hunk 1]
```diff
@@ -27,7 +27,7 @@ int M()
         {
         }
 
-        if ((a = o as string) != null)
+        if ((a = o as string) == null)
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest/UsePatternMatching/CSharpAsAndNullCheckTests_FixAllTests.cs]: ❗️ (Unless not possible) it would be good to retain the original `if` statement as well as the new one.

File: src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs [hunk 2]
```diff
@@ -73,17 +74,42 @@ private void SyntaxNodeAction(SyntaxNodeAnalysisContext syntaxContext)
...
+
+            var comparison = (BinaryExpressionSyntax)leftmostCondition;
+            var operand = GetNullCheckOperand(comparison.Left, comparison.Right)?.WalkDownParentheses();
             if (operand == null)
             {
                 return;
             }
 
+            var semanticModel = syntaxContext.SemanticModel;
+            if (operand.IsKind(SyntaxKind.CastExpression, out CastExpressionSyntax castExpression))
```
Comment (sharwell) [inline src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs]: ❓ Is the ability to handle a cast to `object` new for both the normal and negated forms? If so, we should add a test for the newly supported normal form.

File: src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs [hunk 3]
```diff
@@ -96,8 +122,7 @@ private void SyntaxNodeAction(SyntaxNodeAnalysisContext syntaxContext)
                 return;
             }
 
-            var semanticModel = syntaxContext.SemanticModel;
-            if (semanticModel.GetSymbolInfo(notEquals).GetAnySymbol().IsUserDefinedOperator())
+            if (semanticModel.GetSymbolInfo(comparison).GetAnySymbol().IsUserDefinedOperator())
```
Comment (sharwell) [inline src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs]: 💭 This will rule out the feature for many types. We may be able to improve the experience by allowing user-defined operators, but ensuring a single Fix All operation only corrects operations that use the same operator. For example:

* If all uses of `== null` and `!= null` for a variable use the default operator, a Fix All on the variable will fix all cases where no user-defined operators are involved
* If one or more uses of `==` or `!=` involve a user-defined operator for a variable with the declared type `T`, then a Fix All on that variable will fix all instances involving variables with declared type `T`, whether or not they use user-defined operators. Alternately, this scenario could fix all instances involving variables with declared type `T` which use one or more user-defined operators.

The equivalence ID of the code fix can be used to "bucket" fixes according to the above conditions.

If you like the idea but aren't ready to implement it, feel free to leave this as a follow-up issue by filing a new bug referencing this comment.

File: src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs [hunk 4]
```diff
@@ -210,6 +237,14 @@ private void SyntaxNodeAction(SyntaxNodeAnalysisContext syntaxContext)
                 }
             }
 
+            // If the null check is negative and we're not in a while-statement,
+            // the pattern variable is definitely assigned in the subsequent statements.
+            var isWhile = targetStatement.Kind() == SyntaxKind.WhileStatement;
```
Comment (sharwell) [inline src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs]: ❓ To help me get my head around this, can you provide an example of why this block is necessary?

File: src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs [hunk 5]
```diff
@@ -373,33 +408,36 @@ private static SyntaxNode GetLeftmostCondition(SyntaxNode node)
         {
             while (true)
             {
-                switch (node?.Kind())
```
Comment (sharwell) [inline src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs]: 💭 This seems like an unnecessary change

File: src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs (hunk 3 above)
Comment (alrz) [reply src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs]: This is not part of this PR, the only diff here is `notEquals` -> `comparison`

File: src/EditorFeatures/CSharpTest/UsePatternMatching/CSharpAsAndNullCheckTests_FixAllTests.cs (hunk 1 above)
Comment (alrz) [reply src/EditorFeatures/CSharpTest/UsePatternMatching/CSharpAsAndNullCheckTests_FixAllTests.cs]: I just changed this test to cover negative case.

File: src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs (hunk 2 above)
Comment (alrz) [reply src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs]: Yes, it's new, I'll add the test for `(object)x != null` as well.

File: src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs (hunk 4 above)
Comment (alrz) [reply src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs]:
```cs
if (!(x is Foo  foo))  {} 
// foo is always definitely assigned (we don't need to check so we return early)
while (!(x is Foo  foo)) {}
// foo is always not definitely assigned (so we need to check if it's accessed)
```

File: src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs (hunk 5 above)
Comment (alrz) [reply src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs]: Relates to https://github.com/dotnet/roslyn/issues/21962

I can revert it if you want.

**Comment by alrz** [comment]
Also added fix for https://github.com/dotnet/roslyn/issues/21097

File: src/EditorFeatures/CSharpTest/UsePatternMatching/CSharpAsAndNullCheckTests.cs [hunk 6]
```diff
@@ -20,51 +19,65 @@ internal override (DiagnosticAnalyzer, CodeFixProvider) CreateDiagnosticProvider
...
+            await TestCondition("null == x", "!(o is string x)");
+
+            await TestCondition("(object)x == null", "!(o is string x)");
+            await TestCondition("null == (object)x", "!(o is string x)");
+
+            await TestCondition("(x = o as string) != null", "o is string x");
+            await TestCondition("null != (x = o as string)", "o is string x");
+
+            await TestCondition("(x = o as string) == null", "!(o is string x)");
+            await TestCondition("null == (x = o as string)", "!(o is string x)");
```
Comment (alrz) [inline src/EditorFeatures/CSharpTest/UsePatternMatching/CSharpAsAndNullCheckTests.cs]: is it worth to use `Task.WhenAll` here?

File: src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs (hunk 3 above)
Comment (alrz) [reply src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs]: Filed https://github.com/dotnet/roslyn/issues/23374

**Comment by alrz** [comment]
@sharwell All feedback has been responded. ready for review. Thanks.

File: src/EditorFeatures/CSharpTest/UsePatternMatching/CSharpAsAndNullCheckTests.cs (hunk 6 above)
Comment (CyrusNajmabadi) [reply src/EditorFeatures/CSharpTest/UsePatternMatching/CSharpAsAndNullCheckTests.cs]: Nope. :)

File: src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs (hunk 4 above)
Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs]: can you put this in as a comment.

Comment (alrz) [reply src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs]: Wait, this is only correct if `if` has an unreachable endpoint. I'll remove this check.

**Comment by alrz** [comment]
@CyrusNajmabadi @sharwell for review. Thanks.

This also fixes bugs #21097 and #24286 which are somehow related to whether the check is negative or not.

File: src/EditorFeatures/CSharpTest/UsePatternMatching/CSharpAsAndNullCheckTests.cs [hunk 7]
```diff
@@ -20,51 +19,65 @@ internal override (DiagnosticAnalyzer, CodeFixProvider) CreateDiagnosticProvider
...
         {
-            await TestInRegularAndScriptAsync(
-@"class C
-{
-    void M()
-    {
-        [|var|] x = o as string;
-        if (x != null)
-        {
+            await TestCondition("x != null", "o is string x");
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest/UsePatternMatching/CSharpAsAndNullCheckTests.cs]: :memo: This test case covers the original `InlineTypeCheck1`.

File: src/EditorFeatures/CSharpTest/UsePatternMatching/CSharpAsAndNullCheckTests.cs [hunk 8]
```diff
@@ -20,51 +19,65 @@ internal override (DiagnosticAnalyzer, CodeFixProvider) CreateDiagnosticProvider
...
+            await TestCondition("null != x", "o is string x");
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest/UsePatternMatching/CSharpAsAndNullCheckTests.cs]: :memo: This test case covers the original `InlineTypeCheckInvertedCheck1`.

File: src/EditorFeatures/CSharpTest/UsePatternMatching/CSharpAsAndNullCheckTests.cs [hunk 9]
```diff
@@ -20,51 +19,65 @@ internal override (DiagnosticAnalyzer, CodeFixProvider) CreateDiagnosticProvider
...
+
+            await TestCondition("(object)x != null", "o is string x");
+            await TestCondition("null != (object)x", "o is string x");
+
+            await TestCondition("x == null", "!(o is string x)");
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest/UsePatternMatching/CSharpAsAndNullCheckTests.cs]: :memo: This test case replaces the original `TestMissingOnNullEquality`, showing new functionality.

File: src/EditorFeatures/CSharpTest/UsePatternMatching/CSharpAsAndNullCheckTests.cs [hunk 10]
```diff
@@ -20,51 +19,65 @@ internal override (DiagnosticAnalyzer, CodeFixProvider) CreateDiagnosticProvider
...
+            await TestCondition("null == x", "!(o is string x)");
+
+            await TestCondition("(object)x == null", "!(o is string x)");
+            await TestCondition("null == (object)x", "!(o is string x)");
+
+            await TestCondition("(x = o as string) != null", "o is string x");
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest/UsePatternMatching/CSharpAsAndNullCheckTests.cs]: :memo: This test case covers the original `TestInlineNullCheck1`.

File: src/EditorFeatures/CSharpTest/UsePatternMatching/CSharpAsAndNullCheckTests.cs [hunk 11]
```diff
@@ -20,51 +19,65 @@ internal override (DiagnosticAnalyzer, CodeFixProvider) CreateDiagnosticProvider
...
+            await TestCondition("null != (x = o as string)", "o is string x");
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest/UsePatternMatching/CSharpAsAndNullCheckTests.cs]: :memo: This test case covers the original `TestInlineNullCheck2`.

File: src/EditorFeatures/CSharpTest/UsePatternMatching/CSharpAsAndNullCheckTests.cs [hunk 12]
```diff
@@ -20,51 +19,65 @@ internal override (DiagnosticAnalyzer, CodeFixProvider) CreateDiagnosticProvider
...
-    {
-        if (o is string x)
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsInlineTypeCheck)]
+        public async Task InlineTypeCheck2()
         {
-        }
-    }
-}");
+            await TestCondition("(x = o as string) != null", "o is string x");
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest/UsePatternMatching/CSharpAsAndNullCheckTests.cs]: :memo: This test case covers the original `TestInlineNullCheck3`.

File: src/EditorFeatures/CSharpTest/UsePatternMatching/CSharpAsAndNullCheckTests.cs [hunk 13]
```diff
@@ -20,51 +19,65 @@ internal override (DiagnosticAnalyzer, CodeFixProvider) CreateDiagnosticProvider
...
+            await TestCondition("null != (x = o as string)", "o is string x");
```
Comment (sharwell) [inline src/EditorFeatures/CSharpTest/UsePatternMatching/CSharpAsAndNullCheckTests.cs]: :memo: This test case covers the original `TestInlineNullCheck4`.

File: src/EditorFeatures/CSharpTest/UsePatternMatching/CSharpAsAndNullCheckTests_FixAllTests.cs (hunk 1 above)
Comment (sharwell) [reply src/EditorFeatures/CSharpTest/UsePatternMatching/CSharpAsAndNullCheckTests_FixAllTests.cs]: :arrow_right: Sounds good, thanks

File: src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckCodeFixProvider.cs [hunk 14]
```diff
@@ -77,12 +77,20 @@ private static ExpressionSyntax GetCondition(SyntaxNode node)
             var conditionPart = (BinaryExpressionSyntax)conditionLocation.FindNode(cancellationToken);
             var asExpression = (BinaryExpressionSyntax)asExpressionLocation.FindNode(cancellationToken);
 
-            var updatedConditionPart = SyntaxFactory.IsPatternExpression(
+            var updatedConditionPart = (ExpressionSyntax)SyntaxFactory.IsPatternExpression(
```
Comment (sharwell) [inline src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckCodeFixProvider.cs]: ❗️ Use an explicitly-typed variable rather than a cast. This ensures a compilation error occurs for any case where an implicit cast is not available (which in theory ensures the operation is a widening conversion, though we have cases even in this codebase where that is not the case).

File: src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckCodeFixProvider.cs [hunk 15]
```diff
@@ -77,12 +77,20 @@ private static ExpressionSyntax GetCondition(SyntaxNode node)
...
                 asExpression.Left, SyntaxFactory.DeclarationPattern(
                     ((TypeSyntax)asExpression.Right).WithoutTrivia(),
                     SyntaxFactory.SingleVariableDesignation(
                         localDeclaration.Declaration.Variables[0].Identifier.WithoutTrivia())));
 
+            // We should negate the is-expression if we have something like "x == null"
+            if (conditionPart.Kind() == SyntaxKind.EqualsExpression)
```
Comment (sharwell) [inline src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckCodeFixProvider.cs]: :bulb: `IsKind`

File: src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs [hunk 16]
```diff
@@ -73,17 +74,51 @@ private void SyntaxNodeAction(SyntaxNodeAnalysisContext syntaxContext)
...
+                    isNegativeNullCheck = false;
+                    break;
+                case SyntaxKind.EqualsExpression:
+                    isNegativeNullCheck = true;
+                    break;
+                default:
+                    return;
+            }
+
+            // The pattern variable is only definitly assigned if the pattern succeeded,
```
Comment (sharwell) [inline src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs]: :bulb: definit**e**ly

:bulb: I find this block pleasantly readable, but it may be because I have context from seeing the recent bug. Consider expanding this comment by adding the following:

```csharp
// For example:
//
// if ((x = o as string) != null || SomeExpression) {
//   // x would not be definitely assigned here if pattern matching were used
// }
```

File: src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs [hunk 17]
```diff
@@ -160,13 +194,13 @@ private void SyntaxNodeAction(SyntaxNodeAnalysisContext syntaxContext)
                 additionalLocations));
         }
 
-        private static bool IsAccessedBeforeAssignment(
+        private static bool CheckDefiniteAssignment(
```
Comment (sharwell) [inline src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs]: :bulb:  A comment regarding the return value would be handy

File: src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs [hunk 18]
```diff
@@ -195,71 +232,78 @@ private void SyntaxNodeAction(SyntaxNodeAnalysisContext syntaxContext)
...
-                    return targetStatement.Kind() == SyntaxKind.WhileStatement;
+                    if (isAssigned)
+                    {
+                        return true;
+                    }
+
+                    if (isAccessedBeforeAssignment)
+                    {
+                        // Access before assignment here is only valid if we have a negative
+                        // pattern-matching in an if-statement with an unreachable endpoint.
```
Comment (sharwell) [inline src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs]: 💡 Assuming my understanding is correct, the following might be good to add.

```csharp
// For example:
//
// if (!(o is string x)) {
//   return;
// }
//
// // The 'return' statement above ensures x is definitely assigned here
// Console.WriteLine(x);
```

**Comment by sharwell** [comment]
Requesting a review from @agocke to verify the use of data flow analysis.

Moving back into to 15.6.

File: src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckCodeFixProvider.cs (hunk 15 above)
Comment (alrz) [reply src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckCodeFixProvider.cs]: I use `Kind()` where the code asserts that the variable is never null i.e the path is unreachable if it was. This is also the convention used on the compiler side.

Comment (alrz) [reply src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckCodeFixProvider.cs]: I suppose on the IDE side we traditionally use IsKind to not throw if anything ever failed. I'll update all usages.

**Comment by CyrusNajmabadi** [comment]
> This would be a candidate for using Theory instead of Fact. With that said, Theory is used sparingly (very rarely currently) and not everyone has agreed that it's a good idea.

I'm usually one how pushes back on Theories.  That said, if things stay *very* simple, and it's clear precisely what the theory is testing/expecting, then i'm totally ok with it.  Where i usually don't like them is where people parameterize 6 places in a test and it becomes totally unclear waht's actually happening :)

In this case, theories look fine to me.  If we use them judiciously and keep clarity high, then i'm all for them :)

File: src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs (hunk 18 above)
Comment (alrz) [reply src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs]: Done.

File: src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs [hunk 19]
```diff
@@ -195,71 +240,88 @@ private void SyntaxNodeAction(SyntaxNodeAnalysisContext syntaxContext)
...
+                return;
             }
 
-            isAssigned = false;
-            isAccessedBeforeAssignment = false;
+            var dataFlow = lastStatement is null
+                ? semanticModel.AnalyzeDataFlow(firstStatement)
+                : semanticModel.AnalyzeDataFlow(firstStatement, lastStatement);
+            isAssigned = dataFlow.AlwaysAssigned.Contains(localVariable);
+            isAccessedBeforeAssignment = !isAssigned && dataFlow.ReadInside.Contains(localVariable);
```
Comment (alrz) [inline src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs]: There is a bug here, should be missing but it's not:
```cs
    void M()
    {
        [|var|] x = o as string;
        if (x != null)
        {
        }
        Console.WriteLine(x);
        x = "";
    }
}
```
x is always but there is a read before assignment - probably need to revert code to iterate over statements one by one, @CyrusNajmabadi, can you confirm if that was the original purpose of doing so instead of region analysis?

Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs]: > There is a bug here, should be missing but it's not:

It's missing for me.  I'm not able to see this locally.

Comment (alrz) [reply src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs]: I feel there should be a combination of DataFlowsIn/Out that helps.

Comment (alrz) [reply src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs]: Note: iterating over statements is not safe because assignments could be anywhere in an expression,
```cs
Console.WriteLine( Read(x) ??  (x = "tricky"));
```

Comment (alrz) [reply src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs]: @CyrusNajmabadi I meant in this branch. The previous revision check for nodes statement so doesn't have this bug.

However, with either approaches, this one is not missing.
```cs
    void Use(string x) {}
    void M()
    {
        var x = o as string;
        if (x != null)
        {
        }
        Console.WriteLine(x = Use(x));
    }
```
From data flow analysis results, we can't see if the read is happened before write..

Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs]: >  @CyrusNajmabadi, can you confirm if that was the original purpose of doing so instead of region analysis?

Yes, i believe that was the intent. 

> Note: iterating over statements is not safe because assignments could be anywhere in an expression,

I thought we checked assignment on a per-reference basis.  So we would specifically check the Use(x) and see the problem.  Right?

Comment (alrz) [reply src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs]: I think we return early in the previous code. we see the write first and bail. I'll try a mixed approach, if we are AlwaysAssgined, check if there is a read, if yes, need to walk through tree to see it doesn't come first. early return here could be safe.

Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs]: Oh, i think i see what you're saying.  The logic here is buggy:

https://github.com/dotnet/roslyn/blob/f5afdc19c154e9a93561fa66ec013c2cd7730485/src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs#L230-L233

Specifically, because it sees the ```x =```, it assumes that 'x' is always assigned, and thus the change is safe.  

1. That sucks.  But it's also a very corner case.
2. We could consider fixing it.  Namely that if the reference is the left of an assign, that we have to ensure that the right side is fine as well.  Annoying, but doable.

Comment (alrz) [reply src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs]: Note: using AlwaysAssigned is necessary to handle branches e.g. https://github.com/dotnet/roslyn/issues/21097

Comment (CyrusNajmabadi) [reply src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs]: I see.  The logic here has changed a bit, and i'll admit i don't really fully understand the algorithm as it stands now, so i'm not sure the exact right fix.  But i do think this can be solved.

Also, an efficiency point:

```c#
                    var symbol = semanticModel.GetSymbolInfo(id, cancellationToken).GetAnySymbol();
                    if (localVariable.Equals(symbol))
                    {
```

We should check if the id.ValueText matches the local variable name before bothering to do the binding.  That will avoid a lot of expensive semantic checks.

**Comment by CyrusNajmabadi** [comment]
> Note: using AlwaysAssigned is necessary to handle branches e.g. #21097

BTW, something to consider:

Right now the code is very concept/information dense.  I.e. it's trying to figure out all the cases and encode it all into these variables that it's working with.  You should always feel free to consider decomposing this into the major cases that need to be answered.   i.e.:

```c#
if (ThereIsReadBeforeAssignmentInFollowingStatements(...) ||
    ThereIsSomethingWithTheElseClause(...) ||
    ThereIsSomeOtherCaseThatsAlsoImportant(...))
{
    return;
}
```

Then make each of those approaches clear as to what tehy're checking for and how they do it.  

I'm not saying this will improve things.  But it's something to consider when code starts getting very dense and difficult to understand because it's trying to handle so many cases.

**Comment by CyrusNajmabadi** [comment]
BTW, is there a reason we don't just do:

1.  Look for *cheap* and obvious things that would cause a def assignment error.  If we find any, bail immediately.
2. If we don't find anything, then make the actual change, and actually check the compiler to see if it reported any def assignment issues?  

I thought we already did this approach for some fix providers.  It means not actually having to try to reimplement all the compiler's flow control logic into the IDE.

**Comment by CyrusNajmabadi** [comment]
We already have CSharpInlineDeclarationDiagnosticAnalyzer, which has: 

https://github.com/dotnet/roslyn/blob/408cfe0ab9d88d5c2a89739cb8ad62f595e66784/src/Features/CSharp/Portable/InlineDeclaration/CSharpInlineDeclarationDiagnosticAnalyzer.cs#L250-L256

This is essentially the exact same problem.  We're "inlining" the variable declaration into the pattern.  We should just extract out a helper here that can be used for both analyzers.

**Comment by alrz** [comment]
The diff doesn't help much as I rewrote the def assignment check part. Please review the final file. Thanks.

**Comment by alrz** [comment]
tagging @CyrusNajmabadi

File: src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs [hunk 20]
```diff
@@ -160,106 +200,92 @@ private void SyntaxNodeAction(SyntaxNodeAnalysisContext syntaxContext)
...
-                    cancellationToken);
-
-                if (isAccessedBeforeAssignment)
+                var statement = isNegativeNullCheck ? ifStatement.Statement : ifStatement.Else?.Statement;
+                if (statement != null)
                 {
-                    return true;
-                }
+                    var dataFlow = semanticModel.AnalyzeDataFlow(statement);
+                    if (dataFlow.DataFlowsIn.Contains(localVariable))
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs]: it woudl be helpful to have examples shown in the code of what this is catching.

File: src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs [hunk 21]
```diff
@@ -160,106 +200,92 @@ private void SyntaxNodeAction(SyntaxNodeAnalysisContext syntaxContext)
...
+                    {
+                        return false;
+                    }
 
-                if (isAssigned)
-                {
-                    return false;
+                    if (dataFlow.AlwaysAssigned.Contains(localVariable))
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs]: it woudl be helpful to have examples shown in the code of what this is catching.

**Comment by alrz** [comment]
@dotnet-bot test windows_release_vs-integration_prtest please

**Comment by alrz** [comment]
tagging @sharwell, this is ready to go in with 2 approved reviews.

**Comment by jinujoseph** [comment]
@sharwell  lets take this for 15.7 
@Pilchie for ask mode barcheck before retargetting .

**Comment by alrz** [comment]
FYI, I extracted an extension method from inline decl analyzer and simplified this further.

**Comment by Pilchie** [comment]
Approved for 15.7 - sorry for the delay.  Note that it will need to be re-targeted to the `dev15.7.x` branch.

**Comment by sharwell** [comment]
@alrz This needs to be rebased on the **dev15.7.x** branch. Please let me know if you would like me to do the rebase or if you would like to do it yourself. 😄

**Comment by alrz** [comment]
will rebase when I get home

**Comment by alrz** [comment]
Had to squash commits, it got muddled up in the middle of rebase
